jpasskit/src/main/java/de/brendamour/jpasskit/PKPassBuilder.java

```diff
@@ -0,0 +1,438 @@
+/**
+ * Copyright (C) 2019 Patrice Brend'amour <patrice@brendamour.net>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package de.brendamour.jpasskit;
+
+import de.brendamour.jpasskit.enums.PKPassType;
+import de.brendamour.jpasskit.passes.PKBoardingPass;
+import de.brendamour.jpasskit.passes.PKGenericPass;
+import de.brendamour.jpasskit.passes.PKGenericPassBuilder;
+import de.brendamour.jpasskit.util.BuilderUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+
+import java.awt.Color;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+/**
+ * Allows constructing and validating {@link PKPass} entities.
+ *
+ * @author Igor Stepanov
+ */
+public class PKPassBuilder implements IPKValidateable, IPKBuilder<PKPass> {
+
+    private static final int EXPECTED_AUTHTOKEN_LENGTH = 16;
+
+    private PKPass pkPass;
```

Comment:
I appreciate your suggestion, maybe it's theoretically better. But what is the scenario when current "effective immutability" won't be enough? As per my experience, missing setters is a good reason to consider the whole object immutable and treat it the appropriate way.

I just don't like messing too much with extra fields - as per my experience, that's just extra code, extra source of typo bugs.

Reply:
Also I'm not sure that final fields won't cause problems for deserialization (such as with Jackson).

Reply:
I can accept that effective immutability is enough. (if someone wants to change something, final won't prevent that either).
In the other hand if I am trying too look at this from the users' point of view there are things that make more sense.
When users start to use the api they will try to investigate the classes and I think these users should get an idea about how They can use these classes by just looking at them. For example 

- when I see a final field It tells me that it should never change end of story, but when I see a non-final I either accept that somewhere something might change It or I investigate further... 
- if I see that the fields of a class are protected It makes me think why? Is there any subclasses? should I extend the class?
- etc...

These little things might not make any difference in technical point of view and sometimes they might even add more complexity or additional code here and there but if the result is more easy to use by the users I think It's worth It. After all they want to use a library because they want to make their life easier.

I do not want to argue and I am totally fine with this implementation, just wanted to share my ideas.

Reply:
I wouldn't think that this could cause any problems. I use Jackson with immutable objects and i haven't had any issues so far.